### PR TITLE
PLFM-6462: Add default rule for aborting multipart uploads

### PR DESCRIPTION
**Pull request by marcomarasca**

### CODE REVIEW

**Review by marcomarasca — COMMENTED** [review]
Do not merge. Something is wrong in how we call the S3 API, getting back a response such as: 2 rules found with the same prefix ''. I verified that it is indeed possible to set multiple rules with the same prefix (works fine through the console) so this does not make any sense also given that rules at the bucket level are specified on an empty filter set. I suspect that the S3 client version we are using is serializing the deprecated property "prefix" and the server is not happy with that. In our case we need 2 rules, one for aborting multi-part uploads and another to expire objects, note that even if allowed by the API merging this 2 rules into one does not make sense nor it's feasible (we might have other types of rules and we would have to fix conflicts).

**Comment by marcomarasca** [comment]
My assumption was partially correct: in previous versions of S3 they only had a "prefix" property for a rule. They later on added different types of filters for life cycle rules (e.g. based on object tags) and added a new property that contains a tree of filters instead. For backward compatibility the prefix property was retained, but if a filter is omitted the prefix property is used with an empty string instead to indicate the whole bucket (even when setting it explicitly to null, this might be done server side). Since we wanted to apply the rules to the whole bucket I didn't use any filter and automatically it was setting all the rules using the empty string prefix property. 
Apparently having two rules with the same "prefix" property is not supported by the server. So we need to use the filter instead, the way to specify a filter that applies to the whole bucket is to create an instance of a LifecycleFilter with a null predicate, this was discovered by querying the API on a testing bucket where multiple rules applying to the whole bucket where added from the console.

File: src/main/java/org/sagebionetworks/template/s3/S3BucketBuilderImpl.java [hunk 1]
```diff
@@ -191,44 +196,117 @@ private void configureBucketLifeCycle(S3BucketDescriptor bucket) {
...
 
 		if (bucket.getStorageClassTransitions() != null) {
 			for (S3BucketClassTransition transition : bucket.getStorageClassTransitions()) {
 				String transitionRuleName = transition.getStorageClass().name() + RULE_ID_CLASS_TRANSITION;
 
-				update = addRuleIfNotPresent(rules, transitionRuleName, transitionRuleName, () -> 
-					new Rule().withTransitions(Arrays.asList(
-							new Transition().withStorageClass(transition.getStorageClass()).withDays(transition.getDays())
-					)));
+				update |= addOrUpdateRule(rules, bucket.getName(), transitionRuleName, transition, this::createClassTransitionRule, this::updateClassTransitionRule);
```
Comment (john-hill) [inline src/main/java/org/sagebionetworks/template/s3/S3BucketBuilderImpl.java:209]: if(addRule){
update = true
}

File: src/main/java/org/sagebionetworks/template/s3/S3BucketBuilderImpl.java [hunk 2]
```diff
@@ -191,44 +196,117 @@ private void configureBucketLifeCycle(S3BucketDescriptor bucket) {
...
 	
-	private static boolean addRuleIfNotPresent(List<Rule> rules, String bucket, String ruleName, Supplier<Rule> ruleSupplier) {
-		Optional<Rule> rule = findRule(ruleName, rules);
+	private Rule createAbortMultipartRule(S3BucketDescriptor bucket) {
+		return new Rule()
+				.withAbortIncompleteMultipartUpload(new AbortIncompleteMultipartUpload().withDaysAfterInitiation(ABORT_MULTIPART_UPLOAD_DAYS))
+				.withFilter(allBucketLifecycletFilter());
+	}
+	
+	private boolean updateAbortMultipartRule(Rule rule, S3BucketDescriptor bucket) {
```
Comment (john-hill) [inline src/main/java/org/sagebionetworks/template/s3/S3BucketBuilderImpl.java:236]: could we have a unit test for each method?

**Review by john-hill — APPROVED** [review]

**Pull request merged**